packages/compiler-cli/ngcc/index.ts

```diff
@@ -8,15 +8,15 @@
 import {NodeJSFileSystem, setFileSystem} from '../src/ngtsc/file_system';
 
 import {mainNgcc} from './src/main';
-import {AsyncNgccOptions, NgccOptions, SyncNgccOptions} from './src/ngcc_options';
+import {AsyncNgccOptions, SyncNgccOptions} from './src/ngcc_options';
 
 export {ConsoleLogger} from './src/logging/console_logger';
 export {Logger, LogLevel} from './src/logging/logger';
 export {AsyncNgccOptions, NgccOptions, PathMappings, SyncNgccOptions} from './src/ngcc_options';
 
-export function process(options: AsyncNgccOptions): Promise<void>;
-export function process(options: SyncNgccOptions): void;
-export function process(options: NgccOptions): void|Promise<void> {
+export function process<T extends AsyncNgccOptions|SyncNgccOptions>(options: T):
+    T extends AsyncNgccOptions ? Promise<void>: void;
+export function process(options: AsyncNgccOptions|SyncNgccOptions): void|Promise<void> {
```

Comment:
IMO, this would be more readable as:

```suggestion
export function process<T extends NgccOptions>(options: T):
    T extends AsyncNgccOptions ? Promise<void>: void {
```

Same for `mainNgcc()`. Did I miss something?

Reply:
There are not the same.

https://www.typescriptlang.org/v2/en/play?ts=3.9.0-beta#code/GYVwdgxgLglg9mABAUwB7IiKyAUBKALkQDc4YATAHwAUAnOAWxgGdkAeUigPkQG8AofgF9BaAA5xaURDDDZawAIYRkiAMoBPSADkA5hAgB5MbATM+-RFcSLmWiAH4iSgDatriYaNQSpMucgKyqoAgnY6+kYm8GDmAh629kRQtCCqVl78APRZiADqkgDWzPziktKgkKZIYvQqzMxsACooqNhg5OZh9noGxtXMlJoRfdFmXDhwY7FETYSILWjtnYjdI1EDiA6IdIws7JzkXESHANylPuWIldAxiLVw9cyT08xEaxC9GzGDw5+R-R+80ONHoTFYHDIRwsHloyCgIFoSDQGCwuDw5xE2VyAFFaPRaBdfBVwLcEPc6sgGgB9ZqtZZdcL-UYDIY9AGvCZTAazeaLNrIDqM9ksn5bHZg-aQ7gnKEw6xwhFI1qo7D4TGCIA

Reply:
Hm...it would work for `process()` is `mainNgcc()` was also updated to the same signature (as I suggested above). But I now realize that we would need to make changes in `mainNgcc()` internally to satisfy the return new type.

So, let's keep it as is :grin: